tools/storage/segment.py

```diff
@@ -0,0 +1,303 @@
+#!/usr/bin/env python3
+
+"""
+Copyright 2022 Vectorized, Inc.
+ 
+Licensed as a Redpanda Enterprise file under the Redpanda Community
+License (the "License"); you may not use this file except in compliance with
+the License. You may obtain a copy of the License at
+ 
+https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+"""
+
+import os
+import sys
+import collections
+import struct
+import json
+import crc32c
+import gzip
+import lz4.frame
+import zstd
+import six
+import logging
+
+logger = logging.getLogger(__name__)
+
+"""
+https://docs.python.org/3.9/library/struct.html#format-strings
+
+Redpanda header prefix:
+  - little endian encoded
+  - batch size, base offset, type crc
+
+Note that the crc that is stored is the crc reported by Kafka which happens to
+be computed over the big endian encoding of the same data. thus to verify the
+crc we need to rebuild part of the header in big endian before adding to crc.
+"""
+HDR_FMT_RP_PREFIX_NO_CRC = "iqbI"
+HDR_FMT_RP_PREFIX = "<I" + HDR_FMT_RP_PREFIX_NO_CRC
+
+"""
+Below the crc Redpanda and Kafka have the same layout
+  - little endian encoded
+  - attributes ... record_count
+"""
+HDR_FMT_CRC = "hiqqqhii"
+
+HDR_FMT_RP = HDR_FMT_RP_PREFIX + HDR_FMT_CRC
+HEADER_SIZE = struct.calcsize(HDR_FMT_RP)
+
+Header = collections.namedtuple(
+    'Header', ('header_crc', 'batch_size', 'base_offset', 'type', 'crc',
+               'attrs', 'delta', 'first_ts', 'max_ts', 'producer_id',
+               'producer_epoch', 'base_seq', 'record_count'))
+
+
+def read_varlong(buffer, pos, bits=64):
+    """
+    Decodes zig-zag encoded signed integers up to 64 bits long as per utils/vint.h
+    from `buffer` starting at `pos`.
+    Returns (value, next @pos)
+    """
+    signbit = 1 << (bits - 1)
+    mask = (1 << bits) - 1
+    result = 0
+    shift = 0
+    while 1:
```

Comment:
The bit-twiddling in this function may have unexpectedly long runtime in python compared with when we do it from the C++ side.  It would be very useful to know how long it takes to unpack a 1GB (default size) segment with this tool.

Performance didn't matter here when this was a debug tool, but if it's going into a customer's data pipeline they're likely to care, and may not realize how big segments can get by default when testing.

Reply:
anecdotally the previous version of this tool would decode every batch header and validate checksum. it was quite fast.

@jrkinley you might consider making the decoding of the records which are the only user of variable length encoding optional. that will speed things up, and I would assume that decoding the user payloads is going to be less common.

Reply:
@jrkinley you might consider making the decoding of the records which are the only user of variable length encoding optional. that will speed things up, and I would assume that decoding the user payloads is going to be less common.

fyi: there are some tricks in https://github.com/dpkp/kafka-python/blob/f0a57a6a20a3049dc43fbf7ad9eab9635bd2c0b0/kafka/record/default_records.py for performance related to working on a raw byte buffers.

Reply:
I'll create a 1GB segment and do some testing. Either way, I think it's better to create bindings for the `deserialize` method in `utils/vint.h`.

Reply:
> Either way, I think it's better to create bindings for the deserialize method in utils/vint.h.

cross the language barrier to decode integer is probably slower than doing it in python itself. if we start to hit performance problems we can think about different ways to architect it.